pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -0,0 +1,1577 @@
+		Internal,
+		/// The collateral required for the minting operation exceeds the
+		/// provided maximum cost or the released collateral in the burning
+		/// operation is less than the minimum return.
+		Slippage,
+		/// The calculated collateral is zero.
+		ZeroCollateral,
+	}
+
+	#[pallet::composite_enum]
+	pub enum HoldReason {
+		Deposit,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		<CurveParameterTypeOf<T> as Fixed>::Bits:
+			Copy + ToFixed + AddAssign + BitOrAssign + ShlAssign + TryFrom<U256> + TryInto<U256>,
+		CollateralCurrenciesBalanceOf<T>: Into<U256> + TryFrom<U256>,
+		FungiblesBalanceOf<T>: Into<U256> + TryFrom<U256>,
+		// TODO: make large integer type configurable
+	{
+		/// Creates a new bonded token pool. The pool will be created with the
+		/// given curve, collateral currency, and bonded currencies. The pool
+		/// will be owned by the origin account.
+		///
+		/// # Parameters
+		/// - `origin`: The origin of the call.
+		/// - `curve`: The curve parameters for the pool.
+		/// - `collateral_id`: The ID of the collateral currency.
+		/// - `currencies`: A bounded vector of token metadata for the bonded
+		///   currencies.
+		/// - `denomination`: The denomination for the bonded currencies.
+		/// - `transferable`: A boolean indicating if the bonded currencies are
+		///   transferable.
+		///
+		/// # Returns
+		/// - `DispatchResult`: The result of the dispatch.
+		///
+		/// # Errors
+		/// - `Error::<T>::InvalidInput`: If the denomination is greater than
+		///   the maximum allowed or if the curve input is invalid.
+		/// - `Error::<T>::Internal`: If the conversion to `BoundedVec` fails.
+		/// - Other errors depending on the types in the config.
+		#[pallet::call_index(0)]
+		#[pallet::weight({
+			let currency_length = currencies.len().saturated_into();
+			let weight_polynomial = T::WeightInfo::create_pool_polynomial(currency_length);
+			let weight_square_root = T::WeightInfo::create_pool_square_root(currency_length);
+			let weight_lmsr = T::WeightInfo::create_pool_lmsr(currency_length);
+			weight_polynomial.max(weight_square_root).max(weight_lmsr)
+		})]
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			curve: CurveInput<CurveParameterInputOf<T>>,
+			collateral_id: CollateralAssetIdOf<T>,
+			currencies: BoundedVec<TokenMetaOf<T>, T::MaxCurrencies>,
+			denomination: u8,
+			transferable: bool,
+			min_operation_balance: u128,
+		) -> DispatchResult {
+			let who = T::PoolCreateOrigin::ensure_origin(origin)?;
+
+			ensure!(denomination <= T::MaxDenomination::get(), Error::<T>::InvalidInput);
+			let checked_curve = curve.try_into().map_err(|_| Error::<T>::InvalidInput)?;
+
+			let currency_length = currencies.len();
+
+			let current_asset_id = NextAssetId::<T>::get();
+
+			let (currency_ids, next_asset_id) = Self::generate_sequential_asset_ids(current_asset_id, currency_length)?;
+
+			let pool_id = T::PoolId::from(currency_ids.blake2_256());
+
+			// Todo: change that.
```

Comment:
Is this something we want to tackle still? Or should we remove this?

Reply:
> Is this something we want to tackle still? Or should we remove this?

@Ad96el?

Reply:
I think we can leave it like that